In [1]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load in 

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the "../input/" directory.
# For example, running this (by clicking run or pressing Shift+Enter) will list the files in the input directory

import os
print(os.listdir("../input"))

# Any results you write to the current directory are saved as output.

['ChurnData.csv']


# About Dataset:
We’ll use a telecommunications data for predicting customer churn. This is a historical customer data where each row represents one customer. The data is relatively easy to understand, and you may uncover insights you can use immediately. Typically it’s less expensive to keep customers than acquire new ones, so the focus of this analysis is to predict the customers who will stay with the company.

This data set provides info to help you predict behavior to retain customers. You can analyze all relevant customer data and develop focused customer retention programs.

The data set includes information about:

Customers who left within the last month – the column is called Churn
Services that each customer has signed up for – phone, multiple lines, internet, online security, online backup, device protection, tech support, and streaming TV and movies
Customer account information – how long they’ve been a customer, contract, payment method, paperless billing, monthly charges, and total charges
Demographic info about customers – gender, age range, and if they have partners and dependents

In [2]:
df=pd.read_csv("../input/ChurnData.csv")

In [3]:
df.dropna().head()

tenure   age  address  income   ed  employ  equip  callcard  wireless  \
0    11.0  33.0      7.0   136.0  5.0     5.0    0.0       1.0       1.0   
1    33.0  33.0     12.0    33.0  2.0     0.0    0.0       0.0       0.0   
2    23.0  30.0      9.0    30.0  1.0     2.0    0.0       0.0       0.0   
3    38.0  35.0      5.0    76.0  2.0    10.0    1.0       1.0       1.0   
4     7.0  35.0     14.0    80.0  2.0    15.0    0.0       1.0       0.0   

   longmon  ...  pager  internet  callwait  confer  ebill  loglong  logtoll  \
0     4.40  ...    1.0       0.0       1.0     1.0    0.0    1.482    3.033   
1     9.45  ...    0.0       0.0       0.0     0.0    0.0    2.246    3.240   
2     6.30  ...    0.0       0.0       0.0     1.0    0.0    1.841    3.240   
3     6.05  ...    1.0       1.0       1.0     1.0    1.0    1.800    3.807   
4     7.10  ...    0.0       0.0       1.0     1.0    0.0    1.960    3.091   

   lninc  custcat  churn  
0  4.913      4.0    1.0  
1  3.497      1.0    1.0  
2  3.401      3.0    0.0  
3  4.331      4.0    0.0  
4  4.382      3.0    0.0  

[5 rows x 28 columns]

In [4]:
df["churn"]=df["churn"].astype(int)

In [5]:
df.head()

tenure   age  address  income   ed  employ  equip  callcard  wireless  \
0    11.0  33.0      7.0   136.0  5.0     5.0    0.0       1.0       1.0   
1    33.0  33.0     12.0    33.0  2.0     0.0    0.0       0.0       0.0   
2    23.0  30.0      9.0    30.0  1.0     2.0    0.0       0.0       0.0   
3    38.0  35.0      5.0    76.0  2.0    10.0    1.0       1.0       1.0   
4     7.0  35.0     14.0    80.0  2.0    15.0    0.0       1.0       0.0   

   longmon  ...  pager  internet  callwait  confer  ebill  loglong  logtoll  \
0     4.40  ...    1.0       0.0       1.0     1.0    0.0    1.482    3.033   
1     9.45  ...    0.0       0.0       0.0     0.0    0.0    2.246    3.240   
2     6.30  ...    0.0       0.0       0.0     1.0    0.0    1.841    3.240   
3     6.05  ...    1.0       1.0       1.0     1.0    1.0    1.800    3.807   
4     7.10  ...    0.0       0.0       1.0     1.0    0.0    1.960    3.091   

   lninc  custcat  churn  
0  4.913      4.0      1  
1  3.497      1.0      1  
2  3.401      3.0      0  
3  4.331      4.0      0  
4  4.382      3.0      0  

[5 rows x 28 columns]

In [6]:
X=np.asarray(df[['age','income','tenure']])

In [7]:
y=np.asarray(df["churn"])

In [8]:
from sklearn import preprocessing

In [9]:
X=preprocessing.StandardScaler().fit(X).transform(X)

# Splitting the data for training and testing


In [10]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split( X, y, test_size=0.2, random_state=4)
print ('Train set:', X_train.shape,  y_train.shape)
print ('Test set:', X_test.shape,  y_test.shape)

Train set: (160, 3) (160,)
Test set: (40, 3) (40,)


# Applying Logistic Regression

In [11]:
from sklearn.linear_model import LogisticRegression

In [12]:
LR=LogisticRegression(C=0.1,solver='liblinear').fit(X_train,y_train)

In [13]:
yhat=LR.predict(X_test)

In [14]:
yhat

array([0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0])

# Calculating the probability score for each classification

In [15]:
yhat_prob=LR.predict_proba(X_test)

In [16]:
yhat_prob

array([[0.75767899, 0.24232101],
       [0.75851955, 0.24148045],
       [0.61023128, 0.38976872],
       [0.77691952, 0.22308048],
       [0.65038239, 0.34961761],
       [0.53962364, 0.46037636],
       [0.58652396, 0.41347604],
       [0.69113248, 0.30886752],
       [0.47052009, 0.52947991],
       [0.78928999, 0.21071001],
       [0.82664161, 0.17335839],
       [0.84596246, 0.15403754],
       [0.42237884, 0.57762116],
       [0.50971239, 0.49028761],
       [0.70050159, 0.29949841],
       [0.62561331, 0.37438669],
       [0.73308298, 0.26691702],
       [0.76231387, 0.23768613],
       [0.72133709, 0.27866291],
       [0.59411294, 0.40588706],
       [0.71311578, 0.28688422],
       [0.76639071, 0.23360929],
       [0.83401645, 0.16598355],
       [0.7103537 , 0.2896463 ],
       [0.67852802, 0.32147198],
       [0.84455918, 0.15544082],
       [0.63873024, 0.36126976],
       [0.72727859, 0.27272141],
       [0.51888256, 0.48111744],
       [0.87625921, 0.12374079],
       [0.

# Calculating jaccard score

In [17]:
from sklearn.metrics import jaccard_similarity_score as jcs

In [18]:
score=jcs(yhat,y_test)

/opt/conda/lib/python3.6/site-packages/sklearn/metrics/classification.py:635: DeprecationWarning: jaccard_similarity_score has been deprecated and replaced with jaccard_score. It will be removed in version 0.23. This implementation has surprising behavior for binary and multiclass classification tasks.
  'and multiclass classification tasks.', DeprecationWarning)


In [19]:
print(score)

0.65


# The jaccard score comes out to be 0.65 which means that our model has prediction score of 65% 